File: src/leetcode/p001to100/LeetCode076MinimumWindowSubstring.java
package leetcode.p001to100;

public class LeetCode076MinimumWindowSubstring {
	
  public String minWindow(String s, String t) {
    if (s == null || s.length() == 0) {
      return "";
    }
    if (t == null || t.length() == 0) {
      return "";
    }
    int start = 0;
    int head = 0;
    int deficiency = t.length();
    int minLen = Integer.MAX_VALUE;
    int[] map = new int[128];
    for (int i = 0, tLen = t.length(); i < tLen; i++) {
      map[t.charAt(i)]++;
    }
    for (int end = 0, sLen = s.length(); end < sLen; end++) {
      if (map[s.charAt(end)]-- > 0) {
        deficiency--;
      }
      while (deficiency == 0) {               
        if ((end - start + 1) < minLen) {
          minLen = end - start + 1;
          head = start;
        }
        if (map[s.charAt(start++)]++ == 0) {
          deficiency++;
        }
      }
    }
    return minLen <= s.length() ? s.substring(head, head + minLen) : "";
  }
  
  public static void main(String[] args) {
    // TODO Auto-generated method stub
    
  }

}

File: karakoram/kb/kb0050/src/main/kotlin/com/anonymouscorgi/karakoram/kb0050/LeetCode076MinimumWindowSubstring.kt
package com.anonymouscorgi.karakoram.kb0050

import com.anonymouscorgi.karakoram.annotation.Accepted
import com.anonymouscorgi.karakoram.annotation.Hard

/**
 * LeetCode 76. Minimum Window Substring
 *
 * Given two strings s and t of lengths m and n respectively, return the minimum window
 *
 * of s such that every character in t (including duplicates) is included in the window. If there is no such substring, return the empty string "".
 *
 * The testcases will be generated such that the answer is unique.
 */
@Hard
internal interface LeetCode076MinimumWindowSubstring {

    fun minWindow(s: String, t: String): String

    @Accepted
    object METHOD : LeetCode076MinimumWindowSubstring {

        override fun minWindow(s: String, t: String): String {
            var deficiency = t.length
            val tCountArray = IntArray(128)
            for (char in t) {
                tCountArray[char.code]++
            }

            var minLeft = 0
            var minLength = Int.MAX_VALUE
            var left = 0
            for (right in s.indices) {
                if (tCountArray[s[right].code]-- <= 0) {
                    continue
                }
                --deficiency
                while (deficiency == 0) {
                    if (right - left + 1 < minLength) {
                        minLength = right - left + 1
                        minLeft = left
                    }
                    if (tCountArray[s[left++].code]++ >= 0) {
                        deficiency++
                        break
                    }
                }
            }
            return if (minLength == Int.MAX_VALUE) {
                ""
            } else {
                s.substring(minLeft, minLeft + minLength)
            }
        }
    }
}